Work

ritasumtsova opened the pull request:
Added:

1. Loader to the Departments and Department pages. New components are LoadingSpinner and withLoader HOC.

2. Modal windows to the Departments (add new department, edit a department) and Department (add new employee, edit an employee) pages. New components are ModalWindow, DepartmentForm, EmployeeForm, EditButton. Also changed AddButton component.

didevgen commented on src/HOC/withLoader.jsx:
```diff
@@ -0,0 +1,35 @@
+import React, { Component } from 'react';
+import LoadingSpinner from '../components/LoadingSpinner/LoadingSpinner';
+
+const withLoader = (WrappedComponent) => {
```
fyi 
HOC are considered to be not the best practice anymore:
https://devcore.io/en/react/hocs-vs-hooks-what-to-use-and-why/

didevgen commented on src/components/AddButton/AddButton.jsx:
```diff
@@ -1,29 +1,73 @@
+      body: '',
     };
+
+    this.toggleModalWindow = this.toggleModalWindow.bind(this);
```
no need to bind since it is the arrow function already

didevgen commented on src/components/AddButton/AddButton.jsx:
```diff
@@ -1,29 +1,73 @@
+  openModalWindow() {
+    const { location } = this.props;
+
+    if (location === '/departments') {
+      this.setState({
+        title: 'Add new department',
+        isOpen: true,
+        body: <DepartmentForm />,
+      });
+    } else {
+      this.setState({
+        title: 'Add new employee',
+        isOpen: true,
+        body: <EmployeeForm />,
+      });
+    }
```
What would you do if there was an another flow ? :)

didevgen commented on src/components/Departments/Departments.jsx:
```diff
@@ -47,24 +55,24 @@ export default class Departments extends Component {
       <div>
         <SearchForm />
-        <AddButton />
+        <AddButton text="Add new department +" location={location.pathname} />
```
it is better to use + as icon instead of the raw text

didevgen commented on src/components/EditButton/EditButton.jsx:
```diff
@@ -0,0 +1,68 @@
+  openModalWindow() {
+    const { location } = this.props;
+
+    if (location === '/departments') {
+      this.setState({
+        title: 'Edit department',
+        isOpen: true,
+        body: <DepartmentForm />,
+      });
+    } else {
+      this.setState({
+        title: 'Edit employee',
+        isOpen: true,
+        body: <EmployeeForm />,
+      });
```
same here. The component should be dumb

ritasumtsova commented:
Thanks for your review, I've added font awesome icon for AddButton.
Also changed AddButton and EditButton logic: now I'm passing some needed stuff as props. This helped me to remove conditions inside of the components. And there's no need anymore to use withRouter HOC from React router.

didevgen commented on src/components/EditButton/EditButton.jsx:
```diff
@@ -0,0 +1,52 @@
+          className="EditButton__modal"
+          title={title}
+          isOpen={isOpen}
+          body={body}
```
rename the `body` to something more meaningful
I would also *consider* using this.props.children instead of props.body ... but just a though. Don't like both approaches either

didevgen commented on src/components/EditButton/EditButton.jsx:
```diff
@@ -0,0 +1,52 @@
+        >
+          Edit
+        </Button>
+        <ModalWindow
```
Also please check. If you render 100 buttons. How much dialogs will be created in DOM?

Think about having one dialog component which will render any content inside

didevgen replied:
Similar to that: https://dev.to/zettadam/how-to-handle-multiple-modals-in-a-react-application-2pei

ritasumtsova commented:
I've changed modals logic by adding context to the app. Now we have contexts directory with the ModalContext and ModalContextProvider.
App.js stuff now is wrapped up in ModalContextProvider.
